llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp

```diff
@@ -2255,7 +2255,7 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
     // This is done in other passes, but we want to be able to consume this
     // pattern in InstCombine so we can generate it without creating infinite
     // loops.
-    if (match(Op0, m_Add(m_Value(X), m_Value(Z))) &&
+    if (match(Op0, m_c_Add(m_Value(X), m_Value(Z))) &&
```

Comment:
There no reason to use a commutative match if both operands are `m_Value`

Reply:
The AND statement that follows it has Z in common.

Reply:
It doesn't matter. The `match` function in m_Value always returns true, so the m_Add will be satisfied on the first operand order it tries. It won't ever try the commuted order.

Reply:
The entire `match(Op0, m_c_Add(m_Value(X), m_Value(Z)))` is evaluated before `match(Op1, m_c_Add(m_Value(Y), m_Specific(Z))))` is evaluated. It won't come back try Op0 again if the Op1 match fails.